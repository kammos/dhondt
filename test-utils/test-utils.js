import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { calculatorReducer } from '../src/calculator';


const customRender = (ui, { initialState, store, ...options } = {}) => {
    const actualStore = store ?? configureStore({
        reducer: { calculator: calculatorReducer },
        preloadedState: initialState,
    });

    const Wrapper = ({ children }) => {
        return <Provider store={actualStore}>
            {children}
        </Provider>
    }

    return render(ui, { wrapper: Wrapper, ...options });
}


export * from '@testing-library/react';
export { customRender as render };