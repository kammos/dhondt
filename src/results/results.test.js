/**
 * @jest-environment jsdom
 */
import { render } from 'test-utils';
import Results from '.';

describe("Results component", () => {
    it("should render", () => {
        const initialState = {
            calculator: {
                resultsValid: true,
                results: {
                    '0': {
                        name: 'Red',
                        votes: 100,
                        index: 0,
                        ignoresTreshold: false,
                        seats: 22,
                        difference: 0.2608695652173907
                    },
                    '1': {
                        name: 'Green',
                        votes: 110,
                        index: 1,
                        ignoresTreshold: false,
                        seats: 24,
                        difference: 0.08695652173912904
                    }
                }
            }
        };

        const ui = render(<Results />, { initialState });

        expect(ui.container).toMatchSnapshot();
    });

    it("should render even if there are no results preloaded by state retention", () => {
        const initialState = {
            calculator: {
                seats: 100,
                treshold: 0,
                nextIndex: 4,
                parties: {
                    '0': {
                        name: 'Red',
                        votes: 100,
                        index: 0,
                        ignoresTreshold: false
                    },
                    '1': {
                        name: 'Green',
                        votes: 110,
                        index: 1,
                        ignoresTreshold: false
                    },
                },
                resultsValid: false,
                results: {}
            }
        };

        const ui = render(<Results />, { initialState });

        expect(ui.container).toMatchSnapshot();
    });

    it("should render a proper message if there are no parties in store", () => {
        const initialState = {
            calculator: {
                seats: 100,
                treshold: 0,
                nextIndex: 4,
                parties: {},
                resultsValid: false,
                results: {}
            }
        };

        const ui = render(<Results />, { initialState });

        expect(ui.container).toMatchSnapshot();
    });

    it("should render a proper message if no party lost or gained any seats", () => {
        const initialState = {
            calculator: {
                resultsValid: true,
                results: {
                    '0': {
                        name: 'Red',
                        votes: 100,
                        index: 0,
                        ignoresTreshold: false,
                        seats: 22,
                        difference: 0.0
                    },
                    '1': {
                        name: 'Green',
                        votes: 110,
                        index: 1,
                        ignoresTreshold: false,
                        seats: 24,
                        difference: 0.0
                    }
                }
            }
        };

        const ui = render(<Results />, { initialState });

        expect(ui.container).toMatchSnapshot();
    });
});