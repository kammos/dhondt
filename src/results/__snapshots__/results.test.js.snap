// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results component should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="chartContainer"
    >
      <chartjs-bar>
        {"options":{"plugins":{"title":{"text":"Seats per party","display":true,"font":{"weight":"bold","color":"neutral","size":16}},"datalabels":{"color":"neutral","font":{"weight":"bold","color":"neutral","size":16}}},"responsive":true,"maintainAspectRatio":true,"aspectRatio":1,"x":{"ticks":{"font":{"weight":"bold","color":"neutral","size":16}}}},"data":{"labels":["Green","Red"],"datasets":[{"label":"Seats","data":[24,22],"backgroundColor":["hsl(110, 75%, 60%)","hsl(0, 75%, 60%)"]}]}}
      </chartjs-bar>
    </div>
    <div
      class="chartContainer"
    >
      <chartjs-bar>
        {"options":{"plugins":{"title":{"text":"Difference from ideal number of seats","display":true,"font":{"weight":"bold","size":16}},"datalabels":{"color":"neutral","font":{"weight":"bold","color":"neutral","size":16}}},"responsive":true,"maintainAspectRatio":true,"aspectRatio":1,"indexAxis":"y","scales":{"y":{"ticks":{"font":{"weight":"bold","color":"neutral","size":16}}},"x":{"suggestedMin":-0.5,"suggestedMax":0.5}}},"data":{"labels":["Green","Red"],"datasets":[{"label":"Seats gained or lost due to inaccuracy of D'Hondt method","data":[0.08695652173912904,0.2608695652173907],"backgroundColor":["plus","plus"]}]}}
      </chartjs-bar>
    </div>
    <div
      class="explanationContainer"
    >
      <div
        class="container"
      >
        <header>
          Seems low?
        </header>
        <div
          class="text"
        >
          <p>
            Remember, this simulator only shows resuls for a 
            <em>
              single voting district
            </em>
            . In a real-world election, typically there will be multiple districts, allowing such discrepancies to accumulate.
          </p>
          <p>
            As an example, in the above simulation for a single district, party 
            Red
             
            <strong>
              gained 0.26 seats
            </strong>
            .
          </p>
          <p>
            In a country with 40 voting districts, this would translate to 
            <strong>
              10 seats gained
            </strong>
             nationwide.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Results component should render a proper message if there are no parties in store 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="explanationContainer"
    >
      <div
        class="container"
      >
        <header>
          It seems that you didn't enter any parties
        </header>
        <div
          class="text"
        >
          <p>
            To see election results, please go back to the 
            <a
              class="link"
              href="/setyp"
            >
              Enter data
            </a>
             tab and fill in at least one party.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Results component should render even if there are no results preloaded by state retention 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="chartContainer"
    >
      <chartjs-bar>
        {"options":{"plugins":{"title":{"text":"Seats per party","display":true,"font":{"weight":"bold","color":"neutral","size":16}},"datalabels":{"color":"neutral","font":{"weight":"bold","color":"neutral","size":16}}},"responsive":true,"maintainAspectRatio":true,"aspectRatio":1,"x":{"ticks":{"font":{"weight":"bold","color":"neutral","size":16}}}},"data":{"labels":["Green","Red"],"datasets":[{"label":"Seats","data":[52,48],"backgroundColor":["hsl(110, 75%, 60%)","hsl(0, 75%, 60%)"]}]}}
      </chartjs-bar>
    </div>
    <div
      class="chartContainer"
    >
      <chartjs-bar>
        {"options":{"plugins":{"title":{"text":"Difference from ideal number of seats","display":true,"font":{"weight":"bold","size":16}},"datalabels":{"color":"neutral","font":{"weight":"bold","color":"neutral","size":16}}},"responsive":true,"maintainAspectRatio":true,"aspectRatio":1,"indexAxis":"y","scales":{"y":{"ticks":{"font":{"weight":"bold","color":"neutral","size":16}}},"x":{"suggestedMin":-0.5,"suggestedMax":0.5}}},"data":{"labels":["Green","Red"],"datasets":[{"label":"Seats gained or lost due to inaccuracy of D'Hondt method","data":[-0.3809523809523867,0.3809523809523867],"backgroundColor":["minus","plus"]}]}}
      </chartjs-bar>
    </div>
    <div
      class="explanationContainer"
    >
      <div
        class="container"
      >
        <header>
          Seems low?
        </header>
        <div
          class="text"
        >
          <p>
            Remember, this simulator only shows resuls for a 
            <em>
              single voting district
            </em>
            . In a real-world election, typically there will be multiple districts, allowing such discrepancies to accumulate.
          </p>
          <p>
            As an example, in the above simulation for a single district, party 
            Green
             
            <strong>
              lost 0.38 seats
            </strong>
            .
          </p>
          <p>
            In a country with 40 voting districts, this would translate to 
            <strong>
              15 seats lost
            </strong>
             nationwide.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
